moped-editor/src/App.js

```diff
@@ -22,42 +23,66 @@ const HASURA_ENDPOINT = process.env.REACT_APP_HASURA_ENDPOINT;
 
-const App = () => {
-  const restrictedRoutes = restrictRoutes(routes);
-  const routing = useRoutes(restrictedRoutes);
-  const { user } = useUser();
-
-  // https://www.apollographql.com/docs/react/networking/authentication/#header
-  const httpLink = createHttpLink({ uri: HASURA_ENDPOINT });
+const useClient = (user) =>
+  useMemo(() => {
+    // see: https://www.apollographql.com/docs/react/networking/authentication/#header
+    const httpLink = createHttpLink({ uri: HASURA_ENDPOINT });
 
-  const authLink = setContext((_, { headers }) => {
-    // Get the authentication token and role from user if it exists
-    const token = getJwt(user);
-    const role = getHighestRole(user);
+    const authLink = setContext((_, { headers }) => {
+      // Get the authentication token and role from user if it exists
+      const token = getJwt(user);
+      const role = getHighestRole(user);
 
-    // Return the headers and role to the context so httpLink can read them
-    return {
-      headers: {
-        ...headers,
-        authorization: token ? `Bearer ${token}` : "",
-        "x-hasura-role": role ? role : "",
-      },
-    };
-  });
+      // Return the headers and role to the context so httpLink can read them
+      return {
+        headers: {
+          ...headers,
+          authorization: token ? `Bearer ${token}` : "",
+          "x-hasura-role": role ? role : "",
+        },
+      };
+    });
 
-  const client = new ApolloClient({
-    // Join authLink and httpLink to handle auth in each request
-    link: authLink.concat(httpLink),
-    cache: new InMemoryCache(),
-  });
+    return new ApolloClient({
+      // Join authLink and httpLink to handle auth in each request
+      link: authLink.concat(httpLink),
+      cache: new InMemoryCache({
+        typePolicies: {
+          // a type policy must be added for any type we want to cache that
+          // does not use a column called `id` as the PK
+          project_list_view: {
+            keyFields: ["project_id"],
+          },
+          moped_entity: {
+            keyFields: ["entity_id"],
+          },
+          moped_phases: {
+            keyFields: ["phase_id"],
+          },
+          moped_types: {
+            keyFields: ["type_id"],
+          },
+        },
+      }),
+    });
+  }, [user]);
 
+const App = () => {
+  const [listViewQuery, setListViewQuery] = useState(null);
+  const routing = useRoutes(restrictedRoutes);
+  const { user } = useUser();
+  const client = useClient(user);
   return (
     <ApolloProvider client={client}>
       <StyledEngineProvider injectFirst>
         <LocalizationProvider dateAdapter={AdapterDateFns}>
           <ThemeProvider theme={theme}>
             <GlobalStyles />
-            {routing}
+            <ProjectListViewQueryContext.Provider
+              value={{ listViewQuery, setListViewQuery }}
+            >
+              {routing}
+            </ProjectListViewQueryContext.Provider>
```

Comment:
i ended up creating this context after testing lots of different ways to refetch the project list view query from the ProjectView.

the idea is that the context preserves the most recent project list view `gql` so that it can be refetched. despite what the Apollo docs say, this really seems like the only way to do it—because other methods, e.g., refetching a query by name, require that the query itself is still active in a mounted component.